Scripts/Commands: implement teleport character to NPC

matanshukry opened the pull request:
**Changes proposed:**

-  Added gm command to teleport a character to an npc.

| Syntax | Example |
| ------- | -- |
| `.tele name2 <char_name> id <npc_id>` | `.tele name2 msh id 63847` |
| `.tele name2 <char_name> guid <npc_guid>` | `.tele name2 msh guid 63847` |
| `.tele name2 <char_name> name <npc_name>` | `.tele name2 msh id Incineratus` |

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

Referenced #25870

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] tested in-game

matanshukry committed:
Implemented command to teleport character to npc

Shauren commented on src/server/game/Chat/Chat.h:
```diff
@@ -127,8 +129,10 @@ class TC_GAME_API ChatHandler
+        std::string extractPlayerNameFromLink(std::string const &text);
         // select by arg (name/link) or in-game selection online/offline player or self if a creature is selected
         bool extractPlayerTarget(char* args, Player** player, ObjectGuid* player_guid = nullptr, std::string* player_name = nullptr);
+        bool extractPlayerTarget(std::string const &nameStr, Player** player, ObjectGuid* player_guid = nullptr, std::string* player_name = nullptr);
```
Don't touch this class to add std::string overloads (will be ported from 3.3.5 later)

matanshukry replied:
I didn't really understand what you mean here.
Is the problem with overloads? That is, if I rename the methods is that fine? or should I not create any methods with std::string in the header?

Shauren replied:
This has to do with significant improvements (black-magic like) done to command argument handling done in the other branch (you would be adding code that will definitely be unused later)

Another thing are the issues I mentioned in other comments

(all of current, "old" handling is based on `strtok` and relies on having calls to  `strtok` done)

matanshukry replied:
Right I saw all the strtok handling, I thought making some new methods with std::string and tokenizer is gonna make it better.

I'm still not sure what I can use though; can I just move my methods into the .cpp file and make them non-member methods? is that fine?

Shauren commented on src/server/scripts/Commands/cs_tele.cpp:
```diff
@@ -35,6 +35,9 @@ EndScriptData */
 #include "Player.h"
 #include "RBAC.h"
 #include "WorldSession.h"
+#include <boost/tokenizer.hpp>
```
We have a Tokenizer class in Util.h

matanshukry replied:
Shauren why do we have it / what are the advantages of it?
from quickly I looking I can see one advantage for boost, which is it parses the data on demand as the tokens are accessed rather than the constructor, and it only support multiple characters.

Shauren replied:
Not using boost is the advantage - compile times don't suffer

Generally if we have code for it, it should be preferred over directly using 3rd party libraries

matanshukry replied:
I was talking about parsing that's done in runtime, not compile time.

And I disagree with that; boost isn't just any 3rd party library, it's widely known and used. I'd even argue that if there are any bugs in the code, it's more likely in ours than theirs, since their product is more widely used and tested.
We're just reinventing the wheel here.

But either way I removed it.

Shauren commented on src/server/scripts/Commands/cs_tele.cpp:
```diff
@@ -343,9 +347,203 @@ class tele_commandscript : public CommandScript
+        {
+            case 0: // id
+            {
+                uint32 npcId = std::stol(entry);
```
Use `atoul`, std::stol can throw

Shauren commented on src/server/scripts/Commands/cs_tele.cpp:
```diff
@@ -343,9 +347,203 @@ class tele_commandscript : public CommandScript
+            }
+            case 1: // guid
+            {
+                ObjectGuid::LowType guidLow = std::stoll(entry);
```
Use `atoull`

MaxtorCoder commented on src/server/scripts/Commands/cs_tele.cpp:
```diff
@@ -343,9 +347,203 @@ class tele_commandscript : public CommandScript
+            return false;
+        }
+
+        QueryResult result = WorldDatabase.PQuery("SELECT position_x, position_y, position_z, orientation, map FROM creature %s", whereClause.str().c_str());
```
What if the NPC has a waypoint and isn't on that position. Iirc there's a method to retrieve the Creature/Unit entity dynamically without having to query DB.

matanshukry replied:
To be honest I pretty much just combined 2 existing gm commands, implementation-wise.
That is, the reason I used this it's because it used in the other gm command - `go creature`.

But if you know another way I'm willing to take a look?

MaxtorCoder replied:
Tbf I cannot remember the method that required GUID/Entry MonkaW

Shauren replied:
not ideal but "fine", if its a copy of logic from other commands

matanshukry committed:
Added parenthesis around assignment inside a condition

illfated commented:
PR title suggestion: **Scripts/Commands: implement teleport character to NPC**
(I prefer the present tense instead of past tense, but that is just my personal preference)

jackpoz commented:
As the issue is about 3.3.5, could you please implement it on that branch ? Last PRs that I ported there had a lot of issues/code differences that required manual handling to be ported

matanshukry commented:
@jackpoz I can port it without actually testing it in-game, if that still works?

Jinnaix commented:
> @jackpoz I can port it without actually testing it in-game, if that still works?

You can ask for other people to check out your PR and check ingame functionality since you seem to not have a WotLK client / environment .

Shauren commented on src/server/scripts/Commands/cs_tele.cpp:
```diff
@@ -48,6 +51,7 @@ class tele_commandscript : public CommandScript
             { "add",   rbac::RBAC_PERM_COMMAND_TELE_ADD,   false, &HandleTeleAddCommand,   "" },
             { "del",   rbac::RBAC_PERM_COMMAND_TELE_DEL,    true, &HandleTeleDelCommand,   "" },
             { "name",  rbac::RBAC_PERM_COMMAND_TELE_NAME,   true, &HandleTeleNameCommand,  "" },
+            { "name2",  rbac::RBAC_PERM_COMMAND_TELE_NAME,   true, &HandleTele2Command,  "tele2 name <character> ((npcid <npcid>) | (npcguid <npcguid>))" },
```
Help text should go to world.command table

Shauren commented on src/server/game/Chat/Chat.cpp:
```diff
@@ -964,15 +1000,75 @@ ObjectGuid::LowType ChatHandler::extractLowGuidFromLink(char* text, HighGuid& gu
     // |color|Hplayer:name|h[name]|h|r
-    char* name_str = extractKeyFromLink(text, "Hplayer");
-    if (!name_str)
+    std::string name_str = extractKeyFromLink(text, "Hplayer");
```
This here is still calling `char*     extractKeyFromLink(char* text, char const* linkType, char** something1 = nullptr);` which can return nullptr - assigning it to std::string will crash

matanshukry replied:
Yeah I shouldn't of touched that function, that was a mistake.

Shauren commented on src/server/game/Chat/Chat.cpp:
```diff
@@ -964,15 +1000,75 @@ ObjectGuid::LowType ChatHandler::extractLowGuidFromLink(char* text, HighGuid& gu
+{
+    // |color|Hplayer:name|h[name]|h|r
+    std::string name_str = extractKeyFromLink(text, "Hplayer");
+    normalizePlayerName(name_str);
```
Ignoring errors from `normalizePlayerName`

matanshukry committed:
- Changed back extractPlayerNameFromLink(char*). Shouldn't of touched that
- Check normalizedPlayerName return value
- remove description from code and align the commands
- Use tc tokenizer rather than boost
- using atoul and atoull instead of std::stol and std::stoll (due to exceptions free)

matanshukry committed:
Removed unused typedef

matanshukry committed:
Removed files that weren't suppose to be in there

illfated commented:
Maybe
```diff
  #include "Language.h"
  #include "MapManager.h"
  #include "MotionMaster.h"
+ #include "ObjectAccessor.h"
  #include "ObjectMgr.h"
  #include "PhasingHandler.h"
  #include "Player.h"
```
in `cs_tele.cpp` ?

matanshukry commented:
@illfated are you saying it because of the result from circleci?
I was surprised to see that so I regenerated the proejct with cmake, cleaned all, rebuilt all. "It works on my machine".

I'll wait for the others CI to see what they show, but would love if someone else on master branch can pull and build this as well.

illfated commented:
> @illfated are you saying it because of the result from circleci?

Yes, it is about the result from ci/circleci: **nopch** (nopch build). The pch builds will complete without issues, I presume.

matanshukry committed:
Fixed no-pch error with include

matanshukry commented:
@illfated not completely sure why it's not showing up in the pch version, but like you said that include completed those errors when building with no_pch.
I'll say ahead locally I had other errors as well (on no_pch only), but I didn't see thme on the circleci so I ignored them. I'll fixthen  if they come up too.

illfated commented:
I don't have the necessary experience to explain the coding difference between pch and no_pch, I only know that "pch" means 'pre-compiled headers' and that it refers to compiling with or without PCH.

matanshukry commented:
@illfated I know what it means, I just don't know why one would want to do 2 build versions, one without and with.

illfated commented:
Fair enough. The TC members usually know stuff like that, especially those who handle a lot of PRs. I am thinking more like "why not, if you have the option to check for multiple scenarios".

Shauren commented:
Committed to 3.3.5 d2be584f99a63a165cfd1d047f460a8ebfb87ac3
